### ISSUE TITLE
mrconvert with [] in output does not split gradients

### ISSUE THREAD

**Issue by bjeurissen** [MRtrix3/mrtrix3#1150]
```
mrconvert dwi.mif volume_[].mif
```
results in ``volume_[].mif`` files containing the full gradient table, instead of just a single row of the gradient table.

This can lead to further problems down the road, e.g.:
```
mrcat -axis 3 volume_*.mif all_volumes.mif
```
will now contain a gradient table with n^2 rows, where the original gradient table had n rows.

**Comment by jdtournier** [comment]
Funny, I was just chatting with @dchristiaens about the fact that this syntax won't handle the gradient table... This would probably need to be handled in the [Header::merge()](https://github.com/MRtrix3/mrtrix3/blob/master/core/header.cpp#L39-L67) function, if you want to have a go...?

**Comment by jdtournier** [comment]
Should also consider the PE table while you're at it...

**Comment by bjeurissen** [comment]
Willing to give it a go, but I will need some more pointers...

**Comment by jdtournier** [comment]
OK, to give you a feel for what's going on here, that merge function is there to check and update the output image header with the information contained in each incoming header from the set. I think [this](https://github.com/MRtrix3/mrtrix3/blob/master/core/header.cpp#L64-L66) is where some changes will be required. The DW scheme entries are currently stored as keyval entries. These will be multi-line, one per volume, for the DW scheme and PE entries. At the moment, it only adds the keyval entry from the next image header if not found in the current header, so that needs to change to allow _appending_ to the current entry - but only for entries where that makes sense...

**Comment by bjeurissen** [comment]
Thanks, this is the part that I could grasp. But what about the other way around? In the example above the initial problem is that the gradient info is not split over the files when mrconverting using the []  syntax.

**Comment by jdtournier** [comment]
Ah, crap. Yes, things are a lot more complicated that way. First off, it's going to be very difficult to support edge cases, like two-way splits (e.g. `mrconvert dwi.mif dwi-[]-[].mif`, which splits the data into _slices_), and other cases that are unexpected, but might nonetheless occur (e.g. the DWI file is actually 5D). But if we just shoot for a best-effort basis, and split the DW gradients only if the file is 4D and the split is one level only (so into individual volumes), then the place to handle this would be in the Header::create() call, [within this loop](https://github.com/MRtrix3/mrtrix3/blob/master/core/header.cpp#L263-L269). You should be able to overwrite the header DW_scheme entry with the correct line, just before it gets written to file, before the `(*format_handler)->create(header)` call. You might need to do the same thing as well before line 246 (I think that's the first one to be written). 
To check whether it's OK to split the DW scheme, I'd check that the files being written out are 3D (or the original was 4D), and that the `Pdim` entry is size one (I think that's the variable used to store the indices being split up).

Hope that helps...

**Comment by Lestropie** [comment]
Closed by #1555.
